docs/okeps/okep-5193-user-defined-networks.md

```diff
@@ -0,0 +1,1468 @@
+    // (such as network policy, services, etc). Additionally, IP port security will also be disabled for interfaces attached to this network.
+    // Defaults to `Enabled`.
+    // +optional
+    Mode IPAMMode `json:"mode"`
+    
+    // Lifecycle controls IP addresses management lifecycle.
+    //
+    // The only allowed value is Persistent. When set, OVN Kubernetes assigned IP addresses will be persisted in an
+    // `ipamclaims.k8s.cni.cncf.io` object. These IP addresses will be reused by other pods if requested.
+    // Only supported when "mode" is `Enabled`.
+    //
+    // +optional
+    Lifecycle NetworkIPAMLifecycle `json:"lifecycle,omitempty"`
+}
+
+```
+Suggested API validation rules:
+- `Topology` and `Role` fields are mandatory.
+- `Topology` can be one of `Layer2`, `Layer3`, `Localnet`.
+- `Role` can be one of `Primary`, `Secondary`.
+- `IPAM.Lifecycle` can be `Persistent`.
+- `IPAM.Mode` can be set to `Disabled` only on `Layer2` or `Localnet` topologies for `Secondary` networks, where the `Subnets` parameter must be omitted. When set to `Enabled`, the `Subnets` attribute must be defined.
+- `JoinSubnets` length can be 1 or 2.
+
+#### Cluster scoped CRD
+```golang
+// ClusterUserDefinedNetwork describes shared OVN network across namespaces request.
+type ClusterUserDefinedNetwork struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    // +kubebuilder:validation:Required
+    // +required
+    Spec ClusterUserDefinedNetworkSpec `json:"spec"`
+    // +optional
+    Status ClusterUserDefinedNetworkStatus `json:"status,omitempty"`
+}
+
+// ClusterUserDefinedNetwork defines the desired state of ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetwork struct {
+    // NamespaceSelector Label selector for which namespace network should be available for.
+    // +kubebuilder:validation:Required
+    // +required
+    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector"`
+    
+    // Template is direct specification of UserDefinedNetwork.
+    // +kubebuilder:validation:Required
+    // +required
+    Template *UserDefinedNetworkTemplateSpec `json:"template"`
+}
+
+// UserDefinedNetworkTemplateSpec UserDefinedNetwork spec template.
+type UserDefinedNetworkTemplateSpec struct {
+    // UserDefinedNetworkSpec contains the UserDefinedNetwork specification.
+    Spec UserDefinedNetworkSpec `json:"spec,omitempty"`
+}
+
+// ClusterUserDefinedNetworkStatus contains the observed status of the ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetworkStatus struct {
+    // ActiveNamespaces indicates in which namespaces network is available.
+    ActiveNamespaces []string `json:"activeNamespaces,omitempty"`
+    
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+```
+
+Suggested CRD short-name: `cudn`
+
+#### Example
+Condition reflecting the network readiness:
+```yaml
+status:
+  conditions:
+  - type: NetworkReady
```

Comment:
I feel like status had an API change later on - remember we had to update tests with backwards compatibility - is this the latest?

Reply:
i'll check and update

Reply:
we did, we have  
```
    - lastTransitionTime: "2025-05-13T18:01:05Z"
      message: NetworkAttachmentDefinition has been created
      reason: NetworkAttachmentDefinitionCreated
      status: "True"
      type: NetworkCreated
    - lastTransitionTime: "2025-05-13T18:01:05Z"
      message: Network allocation succeeded for all synced nodes.
      reason: NetworkAllocationSucceeded
      status: "True"
      type: NetworkAllocationSucceeded

```

However we dont yet have the "OVN network created" status yet. I think it is still TODO. I'll clarify it in the enhancement.